### PR TITLE
Add dvicodes

### PR DESCRIPTION
**Pull request by chirsz-ever**
TODO
- [x] add dpx/src/dpx_dvicodes.rs which play the role as [dpx-dvicodes.h](https://github.com/tectonic-typesetting/tectonic/blob/master/tectonic/dpx-dvicodes.h).
- [x] replace the magic numbers in dpx/src/dpx_dvi.rs and other files with the consts.

### CODE REVIEW

**Comment by cormacrelf** [comment]
Might be a good idea to make them all associated constants of an empty struct `DviOpcode`? So they can be namespaced for ad hoc use, but `use DviOpcode::*` in dpx_dvi.rs. Only do it if it helps legibility elsewhere.

**Comment by cormacrelf** [comment]
Reopening to trigger an arxiv run

Edit: Apparently that wasn’t necessary, just wasn’t showing up. Every output is different, there’s a bug in there somewhere,

**Comment by cormacrelf** [comment]
Probably my fault for talking about u8 in #74 — opcodes are multibyte, so try again without converting everything from i32.

**Comment by chirsz-ever** [comment]
> Might be a good idea to make them all associated constants of an empty struct `DviOpcode`? So they can be namespaced for ad hoc use, but `use DviOpcode::*` in dpx_dvi.rs. Only do it if it helps legibility elsewhere.

It should be enough to use `dpx_dvicodes` module.



> Probably my fault for talking about u8 in #74 — opcodes are multibyte, so try again without converting everything from i32.

Reading [the DeVice Independent file format](https://pydvi.readthedocs.io/en/latest/dvi-format.html) some minutes before, I found that opcodes are always one byte, but some opcodes are followed by bytes of paraments:
> The first byte of each command is the operation code, and this code is followed by zero or more bytes that provide parameters to the command.

I will fix my prevoius commit soon.

**Comment by chirsz-ever** [comment]
@cormacrelf  Could you please help me check eff2ec2? I could really not find any fault.
[Here](https://github.com/tectonic-typesetting/tectonic/blob/4dae4e958e13f84363b30b6d63f9c0879d83d9f1/tectonic/dpx-dvi.c#L1749) is the C code corresponding to.

**Comment by chirsz-ever** [comment]
I'm sure the failure test is not caused by me... I just added a file and a mod declaration

**Comment by Mrmaxmeier** [comment]
> I'm sure the failure test is not caused by me...

It seems like the test run is getting stuck due to some too restrictive timeouts in the tester script. I've increased those (Xref https://github.com/Mrmaxmeier/tectonic-on-arXiv/commit/300a3c93692c484f5eaad0b507652e46a7958461).

The change on 1702.06898 is probably some undefined-behavior related issue.

```console
$ ~/_/tectonic-oxidize on oxidize ⨯ target/release/tectonic /tmp/repro/pf_paper.tex
note: this is a BETA release; ask questions and report bugs at https://tectonic.newton.cx/
Running TeX ...
note: warnings were issued by the TeX engine; use --print and/or --keep-logs for details.
Running BibTeX ...
warning: errors were issued by BibTeX, but were ignored; use --print and/or --keep-logs for details.
Rerunning TeX because bibtex was run ...
Rerunning TeX because "pf_paper.aux" changed ...
Running xdvipdfmx ...
Writing /tmp/repro/pf_paper.pdf (1198196 bytes)
Skipped writing 2 intermediate files (use --keep-intermediates to keep them)
$ ~/_/tectonic-oxidize on oxidize ⨯ target/release/tectonic /tmp/repro/pf_paper.tex
note: this is a BETA release; ask questions and report bugs at https://tectonic.newton.cx/
Running TeX ...
note: warnings were issued by the TeX engine; use --print and/or --keep-logs for details.
Running BibTeX ...
warning: errors were issued by BibTeX, but were ignored; use --print and/or --keep-logs for details.
Rerunning TeX because bibtex was run ...
Rerunning TeX because "pf_paper.aux" changed ...
warning: Could not read object from object stream.
warning: Cannot read PDF document catalog. Broken PDF file?
thread 'main' panicked at 'tt_abort', bridge/src/lib.rs:427:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
error: something bad happened inside TeX; its output follows:

===============================================================================
warning: Cannot read PDF document catalog. Broken PDF file?
===============================================================================

error: the TeX engine had an unrecoverable error
caused by: halted on potentially-recoverable error as specified
```

**Comment by burrbull** [comment]
> The change on 1702.06898 is probably some undefined-behavior related issue.

Looks like same with master. I think it due to untested #169

**Comment by chirsz-ever** [comment]
image
It seems that #169 didn't test 1702.06898.

**Comment by chirsz-ever** [comment]
<del>I think there are not other problems except #170.</del>
I need more modesty, caution and test.

**Comment by Mrmaxmeier** [comment]
> ~I think there are not other problems except #170.~
> I need more modesty, caution and test.

Actually the second change seems to be the same problem #170. This PR does not cause this.

**Comment by chirsz-ever** [comment]
> Actually the second change seems to be the same problem #170. This PR does not cause this.

Thanks for comforting. But I still decide to do a binary test.

**Comment by chirsz-ever** [comment]
Sure, the previous failure on 1702.08278 appears accidently.

File: dpx/src/dpx_vf.rs [hunk 1]
```diff
@@ -242,42 +245,43 @@ unsafe fn read_a_font_def(vf_handle: &mut InputHandleWrapper, mut font_id: i32,
...
-                    tt_skip_bytes(4_u32, vf_handle);
-                    if ch_0 < 0x1000000u32 {
-                        read_a_char_def(vf_handle, thisfont, pkt_len, ch_0);
-                    } else {
-                        eprintln!("char={}", ch_0);
-                        panic!("Long character (>24 bits) in VF file.\nI can\'t handle long characters!\n");
-                    }
-                } else if code == 248i32 {
-                    eof = 1i32
+            242 => {
```
Comment (cormacrelf) [inline dpx/src/dpx_vf.rs]: XXX4 here and a few other spots

**Review by cormacrelf — COMMENTED** [review]
Otherwise great

File: dpx/src/dpx_vf.rs (hunk 1 above)
Comment (chirsz-ever) [reply dpx/src/dpx_vf.rs]: This is the [orignal writing](https://github.com/tectonic-typesetting/tectonic/blob/ed2f275faa73f6016ab0ef02a439f4a715186900/tectonic/dpx-vf.c#L218) and I don't know why...
I think the same with you.

**Comment by burrbull** [comment]
LGTM.

**Pull request merged**